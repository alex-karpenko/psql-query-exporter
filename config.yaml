defaults:
  scrape_interval: 30s
  query_timeout: 5s

sources:
  postgres:
    host: localhost
    port: 5432
    user: ${PG_USER}
    password: ${PG_PASSWORD}
    sslmode: require
    metric_prefix: postgres_state

    databases:
      - dbname: postgres
        queries:
          - metric_name: replication_lag
            query: |
              select slot_name, slot_type, active::text, 
              (case when not pg_is_in_recovery() then pg_current_wal_lsn() - restart_lsn end)::float as lag_bytes
              from pg_replication_slots;
            values:
              single:
                field: lag_bytes
                type: float
            var_labels:
              - slot_name
              - slot_type
              - active
