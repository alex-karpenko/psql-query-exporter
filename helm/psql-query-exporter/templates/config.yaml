apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ include "psql-query-exporter.fullname" . }}-config
  labels:
    {{- include "psql-query-exporter.labels" . | nindent 4 }}
data:
  config.yaml: |
{{ toYaml .Values.config | indent 4 }}
