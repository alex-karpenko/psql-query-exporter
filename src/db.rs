use serde::Deserialize;
use std::{fmt::Display, time::Duration};
use tracing::{debug, error};

use openssl::ssl::{SslConnector, SslMethod, SslVerifyMode};
use postgres_openssl::MakeTlsConnector;
use tokio::task::JoinHandle;
use tokio_postgres::{Client, Error, Row};

#[derive(Debug)]
pub struct PostgresConnection {
    db_connection_string: String,
    client: Client,
    connection_handler: JoinHandle<()>,
    sslmode: PostgresSslMode,
    ssl_rootcert: Option<String>,
    default_backoff_interval: Duration,
    max_backoff_interval: Duration,
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum PostgresSslMode {
    Disable,
    Prefer,
    Require,
    #[serde(rename = "verify-ca")]
    VerifyCa,
    #[serde(rename = "verify-full")]
    VerifyFull,
}

impl Default for PostgresSslMode {
    fn default() -> Self {
        Self::Prefer
    }
}

impl Display for PostgresSslMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Self::Disable => "disable",
            Self::Prefer => "prefer",
            Self::Require => "require",
            Self::VerifyCa => "require",
            Self::VerifyFull => "require",
        };
        write!(f, "{s}")
    }
}

impl PostgresSslMode {
    fn get_mode_str(&self) -> &str {
        match self {
            PostgresSslMode::Disable => "Disabled",
            PostgresSslMode::Prefer => "Prefer",
            PostgresSslMode::Require => "Require",
            PostgresSslMode::VerifyCa => "Verify-CA",
            PostgresSslMode::VerifyFull => "Verify-Full",
        }
    }
}

impl PostgresConnection {
    pub async fn new(
        db_connection_string: String,
        sslmode: PostgresSslMode,
        ssl_rootcert: Option<String>,
        default_backoff_interval: Duration,
        max_backoff_interval: Duration,
    ) -> Result<Self, Error> {
        debug!("PostgresConnection::new: construct new postgres connection");

        let mut backoff_interval = default_backoff_interval;

        loop {
            debug!("sslmode={}", sslmode.get_mode_str());
            let mut connector = SslConnector::builder(SslMethod::tls())
                .unwrap_or_else(|e| panic!("unable to create SSL connector: {}", e));

            match sslmode {
                PostgresSslMode::Disable => connector.set_verify(SslVerifyMode::NONE),
                PostgresSslMode::Prefer => connector.set_verify(SslVerifyMode::NONE),
                PostgresSslMode::Require => connector.set_verify(SslVerifyMode::NONE),
                PostgresSslMode::VerifyCa => {
                    connector.set_verify_callback(
                        SslVerifyMode::PEER,
                        |verify_indicator, x509_result| {
                            let allowed_errors: Vec<i32> = vec![
                                openssl_sys::X509_V_ERR_IP_ADDRESS_MISMATCH,
                                openssl_sys::X509_V_ERR_HOSTNAME_MISMATCH,
                                openssl_sys::X509_V_ERR_EMAIL_MISMATCH,
                            ];
                            debug!(
                                "verify_callback, indicator={}, x509_result={}",
                                verify_indicator,
                                x509_result.error()
                            );

                            if !verify_indicator
                                && allowed_errors.contains(&x509_result.error().as_raw())
                            {
                                true
                            } else {
                                verify_indicator
                            }
                        },
                    );
                }
                PostgresSslMode::VerifyFull => connector.set_verify(SslVerifyMode::PEER),
            };

            if let Some(rootcert) = ssl_rootcert.as_ref() {
                debug!("loading CA bundle from {}", rootcert);
                connector
                    .set_ca_file(rootcert)
                    .unwrap_or_else(|e| panic!("unable to load PEM CA file {}: {}", rootcert, e));
            }

            let connector = MakeTlsConnector::new(connector.build());
            let connection = tokio_postgres::connect(&db_connection_string, connector).await;

            match connection {
                Ok((client, connection)) => {
                    let connection_handler = tokio::spawn(async move {
                        debug!("PostgresConnection::new: spawn new connection task");
                        if let Err(e) = connection.await {
                            error!("PostgresConnection: connection closed with error: {}", e);
                        }
                    });

                    return Ok(PostgresConnection {
                        client,
                        db_connection_string,
                        connection_handler,
                        sslmode,
                        ssl_rootcert,
                        default_backoff_interval,
                        max_backoff_interval,
                    });
                }
                Err(e) => {
                    error!("PostgresConnection::new: client error: {e}");
                }
            };

            tokio::time::sleep(backoff_interval).await;
            backoff_interval += default_backoff_interval;
            if backoff_interval > max_backoff_interval {
                backoff_interval = max_backoff_interval;
            }
        }
    }

    async fn reconnect(&mut self) -> Result<&Self, Error> {
        debug!("PostgresConnection::reconnect: try to reconnect");
        let new_connection = PostgresConnection::new(
            self.db_connection_string.clone(),
            self.sslmode.clone(),
            self.ssl_rootcert.clone(),
            self.default_backoff_interval,
            self.max_backoff_interval,
        )
        .await;

        match new_connection {
            Ok(conn) => {
                self.client = conn.client;
                self.connection_handler = conn.connection_handler;
                Ok(self)
            }
            Err(e) => {
                error!("PostgresConnection::reconnect: can't reconnect: {e}");
                Err(e)
            }
        }
    }

    pub async fn query(&mut self, query: &str, query_timeout: Duration) -> Result<Vec<Row>, Error> {
        debug!("PostgresConnection::query: {query:?}");
        let mut backoff_interval = self.default_backoff_interval;

        loop {
            // Set statement timeout
            let result = self
                .client
                .query(
                    format!("set statement_timeout={};", query_timeout.as_millis()).as_str(),
                    &[],
                )
                .await;
            if let Err(e) = result {
                error!("PostgresConnection::query: {e}");
                if e.code().is_none() {
                    debug!("PostgresConnection::query: try to reconnect after error");
                    self.reconnect().await?;
                } else {
                    return Err(e);
                }
            } else {
                // Execute actual query
                let result = self.client.query(query, &[]).await;
                if let Err(e) = result {
                    error!("PostgresConnection::query: {e}");
                    if e.code().is_none() {
                        debug!("PostgresConnection::query: try to reconnect after error");
                        self.reconnect().await?;
                    } else {
                        return Err(e);
                    }
                } else {
                    return result;
                }
            }

            tokio::time::sleep(backoff_interval).await;
            backoff_interval += self.default_backoff_interval;
            if backoff_interval > self.max_backoff_interval {
                backoff_interval = self.max_backoff_interval;
            }
        }
    }
}
