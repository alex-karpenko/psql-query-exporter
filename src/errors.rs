use std::{env, io};
use thiserror::Error;

#[derive(Error)]
pub enum PsqlExporterError {
    #[error("unable to load config file '{}': {}", .filename, .cause)]
    LoadConfigFile { filename: String, cause: io::Error },
    #[error("unable to parse config: {}", .cause.kind)]
    ParseConfigFile {
        #[from]
        cause: figment::Error,
    },
    #[error("unable to substitute environment variable '{}': {}", .variable, .cause)]
    EnvironmentVariableSubstitution {
        variable: String,
        cause: env::VarError,
    },
    #[error("query failed '{}': {}", .query, .cause)]
    PostgresQuery {
        query: String,
        cause: tokio_postgres::Error,
    },
    #[error("unable to create TLS connector: {}", .0)]
    PostgresTlsConnector(openssl::error::ErrorStack),
    #[error("unable to load CA certificate '{}': {}", .rootcert, .cause)]
    PostgresTlsRootCertificate {
        rootcert: String,
        cause: openssl::error::ErrorStack,
    },
    #[error("unable to load client certificate/key '{}': {}", .filename, .cause)]
    PostgresTlsClientCertificate {
        filename: String,
        cause: openssl::error::ErrorStack,
    },
    #[error("TLS client config error: {}", .0)]
    PostgresTlsClientConfig(String),
    #[error("shutdown signal has been received during operation")]
    ShutdownSignalReceived,
    #[error("unable to create metric '{}': {}", .metric, .cause)]
    CreateMetric {
        metric: String,
        cause: prometheus::Error,
    },
    #[error("unable to send task completion status: {}", .0)]
    MetricsBackStatusSend(#[from] tokio::sync::mpsc::error::SendError<usize>),
}

impl std::fmt::Debug for PsqlExporterError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self)
    }
}
