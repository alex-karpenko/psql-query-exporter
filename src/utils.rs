use crate::errors::PsqlExporterError;
use std::{error::Error, time::Duration};
use tokio::{
    select,
    signal::unix::{signal, Signal, SignalKind},
    sync::watch,
};
use tracing::{debug, error, info, instrument};
pub type ShutdownReceiver = watch::Receiver<bool>;
pub type ShutdownSender = watch::Sender<bool>;

#[derive(Debug)]
pub struct SignalHandler {
    terminate: Signal,
    interrupt: Signal,
    quit: Signal,
    hangup: Signal,

    shutdown_channel_tx: ShutdownSender,
    shutdown_channel_rx: ShutdownReceiver,
}

impl SignalHandler {
    pub fn new() -> Result<Self, Box<dyn Error>> {
        let (shutdown_channel_tx, shutdown_channel_rx) = watch::channel(false);
        let receiver = Self {
            terminate: signal(SignalKind::terminate())?,
            interrupt: signal(SignalKind::interrupt())?,
            quit: signal(SignalKind::quit())?,
            hangup: signal(SignalKind::hangup())?,
            shutdown_channel_tx,
            shutdown_channel_rx,
        };

        Ok(receiver)
    }

    pub fn get_rx_channel(&self) -> ShutdownReceiver {
        self.shutdown_channel_rx.clone()
    }

    #[instrument("SignalHandler", skip_all)]
    pub async fn shutdown_on_signal(&mut self) {
        let signal = self.wait_for_signal().await;

        info!(%signal,  "shutting down");
        if let Err(e) = self.shutdown_channel_tx.send(true) {
            error!(error = %e, "can't send shutdown message");
        } else {
            debug!("shutdown message has been sent, waiting until all task stopped");
            self.shutdown_channel_tx.closed().await;
            debug!("shutdown completed");
        }
    }

    async fn wait_for_signal(&mut self) -> &str {
        select! {
            _ = self.terminate.recv() => "TERM",
            _ = self.interrupt.recv() => "INT",
            _ = self.quit.recv() => "QUIT",
            _ = self.hangup.recv() => "HANGUP",
        }
    }
}

pub struct SleepHelper {
    shutdown_channel: ShutdownReceiver,
}

impl SleepHelper {
    pub fn from(shutdown_channel: ShutdownReceiver) -> Self {
        Self { shutdown_channel }
    }

    #[instrument("SleepHandler", skip_all, fields(duration=?sleep_time))]
    pub async fn sleep(&mut self, sleep_time: Duration) -> Result<(), PsqlExporterError> {
        select! {
            _ = self.shutdown_channel.changed() => {
                debug!("shutdown signal has been received");
                return Err(PsqlExporterError::ShutdownSignalReceived);
            },
            _ = tokio::time::sleep(sleep_time) => {
                debug!("timeout has been reached");
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_sleep_normal_timeout() {
        let (_tx, rx) = watch::channel(false);
        let mut helper = SleepHelper::from(rx);
        let sleep_duration = Duration::from_millis(100);

        select! {
            result = helper.sleep(sleep_duration) => {
                assert!(result.is_ok());
            },
            _ = tokio::time::sleep(Duration::from_millis(200)) => {
                panic!("timeout has been reached");
            }
        }
    }

    #[tokio::test]
    async fn test_sleep_shutdown_signal() {
        let (tx, rx) = watch::channel(false);
        let mut helper = SleepHelper::from(rx);
        let sleep_duration = Duration::from_secs(1);

        // Send shutdown signal
        tokio::spawn(async move {
            tokio::time::sleep(Duration::from_millis(50)).await;
            tx.send(true).unwrap();
        });

        let result = helper.sleep(sleep_duration).await;
        assert!(matches!(
            result,
            Err(PsqlExporterError::ShutdownSignalReceived)
        ));
    }
}
