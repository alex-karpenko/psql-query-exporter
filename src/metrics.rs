use crate::config::{
    FieldType, ScrapeConfig, ScrapeConfigDatabase, ScrapeConfigQuery, ScrapeConfigValues,
};
use crate::db::{PostgresConnection, PostgresSslCertificates};
use crate::errors::PsqlExporterError;
use crate::utils::{ShutdownReceiver, SleepHelper};
use human_repr::HumanDuration;
use prometheus::core::{AtomicF64, AtomicI64, Collector, GenericGauge, GenericGaugeVec};
use prometheus::{
    opts, Encoder, Gauge, GaugeVec, IntGauge, IntGaugeVec, Opts, Registry, TextEncoder,
};
use std::collections::HashMap;
use std::time::{Duration, SystemTime};
use tokio::sync::mpsc;
use tokio_postgres::Row;
use tracing::{debug, error, info, instrument, warn};

#[derive(Debug)]
pub enum MetricWithType {
    SingleInt(GenericGauge<AtomicI64>),
    SingleFloat(GenericGauge<AtomicF64>),
    VectorInt(GenericGaugeVec<AtomicI64>),
    VectorFloat(GenericGaugeVec<AtomicF64>),
}

impl MetricWithType {
    fn to_collector(&self) -> Box<dyn Collector> {
        match self {
            MetricWithType::SingleInt(m) => Box::new(m.to_owned()),
            MetricWithType::SingleFloat(m) => Box::new(m.to_owned()),
            MetricWithType::VectorInt(m) => Box::new(m.to_owned()),
            MetricWithType::VectorFloat(m) => Box::new(m.to_owned()),
        }
    }
}

struct QueryMetrics {
    metrics: Vec<MetricWithType>,
    is_registered: bool,
    last_updated: SystemTime,
    next_query_time: SystemTime,
}

impl QueryMetrics {
    fn from(query_config: &ScrapeConfigQuery) -> Result<Self, PsqlExporterError> {
        let mut metrics: Vec<MetricWithType> = vec![];

        match &query_config.values {
            ScrapeConfigValues::ValueFrom { single: values } => {
                let mut opts = opts!(
                    query_config.metric_name.clone(),
                    query_config.description.clone().unwrap()
                );

                if let Some(const_labels) = &query_config.const_labels {
                    let const_labels: HashMap<String, String> =
                        const_labels.clone().into_iter().collect();
                    opts = opts.const_labels(const_labels);
                }

                let new_metric =
                    Self::helper_create_metric(&query_config.var_labels, &values.field_type, opts)
                        .map_err(|e| PsqlExporterError::CreateMetric {
                            metric: query_config.metric_name.clone(),
                            cause: e,
                        })?;

                metrics.push(new_metric);
            }

            ScrapeConfigValues::ValuesWithLabels {
                multi_labels: values,
            } => {
                for value in values {
                    let mut opts = opts!(
                        query_config.metric_name.clone(),
                        query_config.description.clone().unwrap()
                    );

                    let mut labels = HashMap::new();
                    value.labels.iter().for_each(|(k, v)| {
                        labels.insert(k.to_string(), v.to_string());
                    });

                    if let Some(const_labels) = &query_config.const_labels {
                        const_labels.iter().for_each(|(k, v)| {
                            labels.insert(k.to_string(), v.to_string());
                        });
                    }

                    opts = opts.const_labels(labels);
                    let new_metric = Self::helper_create_metric(
                        &query_config.var_labels,
                        &value.field_type,
                        opts,
                    )
                    .map_err(|e| PsqlExporterError::CreateMetric {
                        metric: query_config.metric_name.clone(),
                        cause: e,
                    })?;

                    metrics.push(new_metric);
                }
            }

            ScrapeConfigValues::ValuesWithSuffixes {
                multi_suffixes: values,
            } => {
                for value in values {
                    let metric_name = format!("{}_{}", query_config.metric_name, value.suffix);
                    let metric_desc = format!(
                        "{}: {}",
                        query_config.description.clone().unwrap(),
                        value.suffix
                    );
                    let mut opts = opts!(metric_name, metric_desc);

                    if let Some(const_labels) = &query_config.const_labels {
                        let const_labels: HashMap<String, String> =
                            const_labels.clone().into_iter().collect();
                        opts = opts.const_labels(const_labels);
                    }
                    let new_metric = Self::helper_create_metric(
                        &query_config.var_labels,
                        &value.field_type,
                        opts,
                    )
                    .map_err(|e| PsqlExporterError::CreateMetric {
                        metric: query_config.metric_name.clone(),
                        cause: e,
                    })?;

                    metrics.push(new_metric);
                }
            }
        };

        Ok(QueryMetrics {
            metrics,
            is_registered: false,
            last_updated: SystemTime::now() - query_config.metric_expiration_time,
            next_query_time: SystemTime::now(),
        })
    }

    fn helper_create_metric(
        var_labels: &Option<Vec<String>>,
        field_type: &FieldType,
        opts: Opts,
    ) -> Result<MetricWithType, prometheus::Error> {
        if let Some(var_labels) = var_labels {
            let new_labels: Vec<&str> = var_labels.iter().map(AsRef::as_ref).collect();
            match field_type {
                FieldType::Int => Ok(MetricWithType::VectorInt(IntGaugeVec::new(
                    opts,
                    &new_labels,
                )?)),
                FieldType::Float => Ok(MetricWithType::VectorFloat(GaugeVec::new(
                    opts,
                    &new_labels,
                )?)),
            }
        } else {
            match field_type {
                FieldType::Int => Ok(MetricWithType::SingleInt(IntGauge::with_opts(opts)?)),
                FieldType::Float => Ok(MetricWithType::SingleFloat(Gauge::with_opts(opts)?)),
            }
        }
    }

    fn register(&mut self, registry: &Registry) {
        self.last_updated = SystemTime::now();
        if !self.is_registered {
            for metric in self.metrics.iter() {
                let boxed_metric = metric.to_collector();
                if let Err(e) = registry.register(boxed_metric) {
                    error!(?metric, "error while registering metric: {e}");
                    return;
                } else {
                    debug!(?metric, "metric registered");
                }
            }
            self.is_registered = true;
        }
    }

    fn unregister(&mut self, registry: &Registry) {
        if self.is_registered {
            for metric in self.metrics.iter() {
                let boxed_metric = metric.to_collector();
                if let Err(e) = registry.unregister(boxed_metric) {
                    error!(?metric, "error while unregistering metric: {e}");
                    return;
                } else {
                    debug!(?metric, "metric unregistered");
                }
            }
            self.is_registered = false;
        }
    }
}

#[instrument("ComposeReply")]
pub async fn compose_reply(registry: Registry) -> String {
    debug!(?registry, "preparing metrics");

    let mut buffer = vec![];
    let encoder = TextEncoder::new();
    let metric_families = registry.gather();
    encoder
        .encode(&metric_families, &mut buffer)
        .unwrap_or_else(|e| panic!("looks like a BUG: {e}"));

    if buffer.is_empty() {
        warn!("no metrics found");
        return String::from("# no metrics found\n");
    }

    String::from_utf8(buffer).unwrap_or_else(|e| panic!("looks like a BUG: {e}"))
}

#[instrument("CollectorsTask", skip_all)]
pub async fn collectors_task(
    scrape_config: ScrapeConfig,
    registry: Registry,
    shutdown_channel: ShutdownReceiver,
) -> Result<(), PsqlExporterError> {
    debug!(config = ?scrape_config);

    if scrape_config.is_empty() {
        warn!("no sources configured, waiting for shutdown signal");
        let mut rx = shutdown_channel.clone();
        rx.changed()
            .await
            .map_err(|_| PsqlExporterError::ShutdownSignalReceived)?;
    } else {
        let mut handler_index: usize = 0;
        let (tx, mut rx) = mpsc::channel(scrape_config.len());
        let sources = scrape_config.sources;
        for (_, source_db_instance) in sources {
            let databases = source_db_instance.databases;
            for database in databases {
                let tx = tx.clone();
                let shut_rx = shutdown_channel.clone();
                let registry = registry.clone();
                tokio::spawn(async move {
                    let handler_result = collect_one_db_instance(database, registry, shut_rx).await;
                    let send_result = tx
                        .send(handler_index)
                        .await
                        .map_err(PsqlExporterError::MetricsBackStatusSend);

                    if let Err(result) = handler_result {
                        match result {
                            PsqlExporterError::ShutdownSignalReceived => {
                                debug!(task = %handler_index, "completed due to shutdown signal");
                                Ok(())
                            }
                            _ => {
                                error!(task = %handler_index, error=%result, "completed unexpectedly");
                                Err(result)
                            }
                        }
                    } else if let Err(result) = send_result {
                        Err(result)
                    } else {
                        handler_result
                    }
                });
                handler_index += 1;
            }
        }

        debug!(task = %handler_index, "handlers have been started");

        while let Some(task_index) = rx.recv().await {
            debug!(task = %task_index, "completed");
            handler_index -= 1;
            if handler_index == 0 {
                info!("all tasks have been stopped, exiting");
                return Ok(());
            }
        }
    }

    Ok(())
}

#[instrument("CollectSingleDbInstance", skip_all, fields(%database))]
async fn collect_one_db_instance(
    database: ScrapeConfigDatabase,
    registry: Registry,
    shutdown_channel: ShutdownReceiver,
) -> Result<(), PsqlExporterError> {
    if database.queries.is_empty() {
        warn!("no queries configured, exiting");
        return Ok(());
    }
    debug!("start task");

    let certificates =
        PostgresSslCertificates::from(database.sslrootcert, database.sslcert, database.sslkey)?;
    let mut db_connection = PostgresConnection::new(
        database.connection_string,
        database.sslmode.unwrap(),
        certificates,
        database.backoff_interval,
        database.max_backoff_interval,
        shutdown_channel.clone(),
    )
    .await?;

    let mut query_metrics: Vec<QueryMetrics> = Vec::with_capacity(database.queries.len());
    let mut sleeper = SleepHelper::from(shutdown_channel.clone());

    for q in database.queries.iter() {
        let metric = QueryMetrics::from(q)?;
        query_metrics.push(metric);
    }

    loop {
        for (query_item, index) in database.queries.iter().zip(0..query_metrics.len()) {
            if query_metrics[index].next_query_time > SystemTime::now() {
                continue;
            }

            let result = db_connection
                .query(&query_item.query, query_item.query_timeout)
                .await;

            match result {
                Ok(result) => {
                    query_metrics[index].register(&registry);
                    let update_result = match &query_item.values {
                        ScrapeConfigValues::ValueFrom { single: value } => {
                            if let Some(field) = &value.field {
                                update_metrics(
                                    &result,
                                    Some(field),
                                    &query_item.var_labels,
                                    &query_metrics[index].metrics[0],
                                )
                            } else {
                                update_metrics(
                                    &result,
                                    None,
                                    &query_item.var_labels,
                                    &query_metrics[index].metrics[0],
                                )
                            }
                        }
                        ScrapeConfigValues::ValuesWithLabels {
                            multi_labels: values,
                        } => {
                            let mut r = Ok(());
                            for (value, metric) in values.iter().zip(&query_metrics[index].metrics)
                            {
                                if let Err(e) = update_metrics(
                                    &result,
                                    Some(&value.field),
                                    &query_item.var_labels,
                                    metric,
                                ) {
                                    r = Err(e);
                                    break;
                                }
                            }
                            r
                        }
                        ScrapeConfigValues::ValuesWithSuffixes {
                            multi_suffixes: values,
                        } => {
                            let mut r = Ok(());
                            for (value, metric) in values.iter().zip(&query_metrics[index].metrics)
                            {
                                if let Err(e) = update_metrics(
                                    &result,
                                    Some(&value.field),
                                    &query_item.var_labels,
                                    metric,
                                ) {
                                    r = Err(e);
                                    break;
                                }
                            }
                            r
                        }
                    };
                    if let Err(e) = update_result {
                        error!("{e}")
                    }
                }
                Err(e) => {
                    if query_item.metric_expiration_time != Duration::ZERO {
                        let expiration_time =
                            query_metrics[index].last_updated + query_item.metric_expiration_time;
                        if SystemTime::now() > expiration_time {
                            debug!("deregister expired metrics");
                            query_metrics[index].unregister(&registry);
                        }
                    }
                    error!("{e}")
                }
            };
            query_metrics[index].next_query_time = SystemTime::now() + query_item.scrape_interval;
        }

        let next_query_time = query_metrics
            .iter()
            .min_by(|x, y| x.next_query_time.cmp(&y.next_query_time))
            .map(|x| x.next_query_time)
            .expect("looks like a BUG");

        let sleep_time;

        if next_query_time > SystemTime::now() {
            sleep_time = next_query_time
                .duration_since(SystemTime::now())
                .unwrap_or(Duration::from_micros(0));
        } else {
            sleep_time = Duration::from_micros(0);

            let slip_duration = SystemTime::now().duration_since(next_query_time).unwrap();
            let slip_duration = slip_duration.human_duration();
            warn!(sleep = %slip_duration, "overtimed query loop");
        }

        sleeper.sleep(sleep_time).await?;
    }
}

#[instrument("UpdateMetrics", skip_all)]
fn update_metrics(
    rows: &[Row],
    field: Option<&str>,
    var_labels: &Option<Vec<String>>,
    metric: &MetricWithType,
) -> Result<(), PsqlExporterError> {
    debug!(?rows, ?field, ?var_labels, ?metric);

    match metric {
        MetricWithType::SingleInt(metric) => {
            if let Some(field) = field {
                metric.set(rows[0].try_get(field)?);
            } else {
                metric.set(rows[0].try_get(0)?);
            }
        }
        MetricWithType::SingleFloat(metric) => {
            if let Some(field) = field {
                metric.set(rows[0].try_get(field)?)
            } else {
                metric.set(rows[0].try_get(0)?)
            }
        }
        MetricWithType::VectorInt(metric) => {
            for row in rows {
                let mut new_labels: Vec<String> = vec![];
                if let Some(labels) = var_labels {
                    for label in labels {
                        new_labels.push(row.try_get(label.as_str())?);
                    }
                    let new_labels: Vec<&str> = new_labels.iter().map(AsRef::as_ref).collect();
                    let new_labels: &[&str] = new_labels.as_slice();
                    if let Some(field) = field {
                        metric
                            .with_label_values(new_labels)
                            .set(row.try_get(field)?);
                    } else {
                        metric.with_label_values(new_labels).set(row.try_get(0)?);
                    }
                }
            }
        }
        MetricWithType::VectorFloat(metric) => {
            for row in rows {
                let mut new_labels: Vec<String> = vec![];
                if let Some(labels) = var_labels {
                    for label in labels {
                        new_labels.push(row.try_get(label.as_str())?);
                    }
                    let new_labels: Vec<&str> = new_labels.iter().map(AsRef::as_ref).collect();
                    let new_labels: &[&str] = new_labels.as_slice();
                    if let Some(field) = field {
                        metric
                            .with_label_values(new_labels)
                            .set(row.try_get(field)?);
                    } else {
                        metric.with_label_values(new_labels).set(row.try_get(0)?);
                    }
                }
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        db::PostgresSslMode,
        test_utils::{
            create_test_connection_string, init_psql_server, init_tracing, TEST_DB_PASSWORD,
            TEST_DB_USER,
        },
    };
    use insta::assert_snapshot;
    use prometheus::Registry;
    use rstest::rstest;
    use std::env;
    use tokio::sync::watch;

    #[rstest]
    #[case("single", 2)]
    #[case("multi_labels", 2)]
    #[case("multi_suffixes", 2)]
    #[tokio::test]
    async fn test_collect_one_db_instance_single_basic(
        #[case] case_name: &str,
        #[case] number_of_updates: usize,
    ) {
        use insta::with_settings;
        use tokio::fs;

        init_tracing().await;
        let port = init_psql_server().await;

        let registry = Registry::new();
        let (tx, rx) = watch::channel(false);

        // Configure collectors task
        env::set_var("TEST_PG_PORT", format!("{port}"));
        env::set_var("TEST_PG_USER", TEST_DB_USER);
        env::set_var("TEST_PG_PASSWORD", TEST_DB_PASSWORD);

        let config =
            ScrapeConfig::from_file(&format!("tests/cases/{case_name}/config.yaml")).unwrap();
        let handler = tokio::spawn(collectors_task(config, registry.clone(), rx.clone()));
        tokio::time::sleep(Duration::from_secs(1)).await;

        // Create side db connection to push updates
        let connection_string = create_test_connection_string(PostgresSslMode::Disable).await;
        let mut db = PostgresConnection::new(
            connection_string,
            PostgresSslMode::Disable,
            PostgresSslCertificates::from(None, None, None).unwrap(),
            Duration::from_secs(1),
            Duration::from_secs(5),
            rx,
        )
        .await
        .unwrap();

        // Get original data
        let metrics = compose_reply(registry.clone()).await;
        with_settings!(
            { description => format!("collector test case '{}', original data", case_name), omit_expression => true },
            { assert_snapshot!(format!("{case_name}_original"), metrics) }
        );

        // Update data
        for round in 1..=number_of_updates {
            // get queries string from file
            let queries = fs::read_to_string(format!("tests/cases/{case_name}/update_{round}.sql"))
                .await
                .unwrap();
            // Split it and filter out empty ones and comments
            let queries: Vec<&str> = queries
                .split('\n')
                .map(|s| s.trim())
                .filter(|s| !s.is_empty())
                .filter(|s| !s.starts_with("--"))
                .collect();
            // Run queries one by one
            for query in queries {
                db.query(query, Duration::from_secs(1)).await.unwrap();
            }
            // Wait for more than 2s for the next scrape round
            tokio::time::sleep(Duration::from_secs(3)).await;
            let metrics = compose_reply(registry.clone()).await;
            with_settings!(
                { description => format!("collector test case '{case_name}', after update {round}"), omit_expression => true },
                { assert_snapshot!(format!("{case_name}_updated_{round}"), metrics) }
            );
        }

        // Wait for expiration
        tokio::time::sleep(Duration::from_secs(6)).await;
        let metrics = compose_reply(registry).await;
        with_settings!(
            { description => format!("collector test case '{}', expired", case_name), omit_expression => true },
            { assert_snapshot!(format!("{case_name}_expired"), metrics) }
        );

        tx.send(true).unwrap();
        handler.await.unwrap().unwrap();
    }
}
