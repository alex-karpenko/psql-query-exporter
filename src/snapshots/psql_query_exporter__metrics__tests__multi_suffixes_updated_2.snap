---
source: src/metrics.rs
description: "collector test case 'multi_suffixes', after update 2"
---
# HELP multi_suffixes_with_const_labels_float_suffix multi_suffixes_with_const_labels: float_suffix
# TYPE multi_suffixes_with_const_labels_float_suffix gauge
multi_suffixes_with_const_labels_float_suffix{const_label_key_1="const_label_value_1",const_label_key_2="const_label_value_2"} 66.66
# HELP multi_suffixes_with_const_labels_int_suffix multi_suffixes_with_const_labels: int_suffix
# TYPE multi_suffixes_with_const_labels_int_suffix gauge
multi_suffixes_with_const_labels_int_suffix{const_label_key_1="const_label_value_1",const_label_key_2="const_label_value_2"} 66
# HELP multi_suffixes_with_expiration_count multi_suffixes_with_expiration: count
# TYPE multi_suffixes_with_expiration_count gauge
multi_suffixes_with_expiration_count 6
# HELP multi_suffixes_with_var_labels_count multi_suffixes_with_var_labels: count
# TYPE multi_suffixes_with_var_labels_count gauge
multi_suffixes_with_var_labels_count{name="Jack"} 1
multi_suffixes_with_var_labels_count{name="Jane"} 1
multi_suffixes_with_var_labels_count{name="John"} 1
# HELP multi_suffixes_with_var_labels_float_suffix multi_suffixes_with_var_labels: float_suffix
# TYPE multi_suffixes_with_var_labels_float_suffix gauge
multi_suffixes_with_var_labels_float_suffix{name="Jack"} 23.32
multi_suffixes_with_var_labels_float_suffix{name="Jane"} 22.22
multi_suffixes_with_var_labels_float_suffix{name="John"} 21.12
# HELP multi_suffixes_with_var_labels_int_suffix multi_suffixes_with_var_labels: int_suffix
# TYPE multi_suffixes_with_var_labels_int_suffix gauge
multi_suffixes_with_var_labels_int_suffix{name="Jack"} 23
multi_suffixes_with_var_labels_int_suffix{name="Jane"} 22
multi_suffixes_with_var_labels_int_suffix{name="John"} 21
# HELP multi_suffixes_without_common_labels_float_suffix multi_suffixes_without_common_labels: float_suffix
# TYPE multi_suffixes_without_common_labels_float_suffix gauge
multi_suffixes_without_common_labels_float_suffix 66.66
# HELP multi_suffixes_without_common_labels_int_suffix multi_suffixes_without_common_labels: int_suffix
# TYPE multi_suffixes_without_common_labels_int_suffix gauge
multi_suffixes_without_common_labels_int_suffix 66
# HELP multi_suffixes_without_common_labels_with_desc_float_suffix single basic with description: float_suffix
# TYPE multi_suffixes_without_common_labels_with_desc_float_suffix gauge
multi_suffixes_without_common_labels_with_desc_float_suffix 66.66
# HELP multi_suffixes_without_common_labels_with_desc_int_suffix single basic with description: int_suffix
# TYPE multi_suffixes_without_common_labels_with_desc_int_suffix gauge
multi_suffixes_without_common_labels_with_desc_int_suffix 66
